### PR TITLE
For #6225: Make webchannel work with custom fxa server

### PR DESCRIPTION
**Pull request by jackyzy823**
Make webchannel work with custom fxa server via background message, for #6225 .

finally found out that  `<all_urls>` permission is required to make https://github.com/mozilla-mobile/android-components/issues/6225#issuecomment-637585738 work.

### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

**Comment by jackyzy823** [comment]
:disappointed:  Could anyone review this? Thanks.

**Comment by jackyzy823** [comment]
ping @grigoryk

**Comment by jackyzy823** [comment]
@csadilek could you help reviewing this since you were in #6225  discussion. Thanks.

File: components/feature/accounts/src/main/assets/extensions/fxawebchannel/manifest.template.json [hunk 1]
```diff
@@ -20,10 +20,14 @@
...
+  "background": {
+    "scripts": ["background.js"]
+  },
   "permissions": [
     "mozillaAddons",
     "geckoViewAddons",
     "nativeMessaging",
-    "nativeMessagingFromContent"
+    "nativeMessagingFromContent",
+    "<all_urls>"
```
[the remark this answers is not in the document]
Comment (jackyzy823) [reply components/feature/accounts/src/main/assets/extensions/fxawebchannel/manifest.template.json:31]: Seems no. I tested `activeTab`  which results endless loading icon just like before.

File: components/feature/accounts/src/main/java/mozilla/components/feature/accounts/FxaWebChannelFeature.kt [hunk 2]
```diff
@@ -167,12 +170,21 @@ class FxaWebChannelFeature(
         extensionController.registerContentMessageHandler(engineSession, messageHandler)
     }
 
+    private class WebChannelViewBackgroundMessageHandler(
+        private val serverConfig: ServerConfig
+    ) : MessageHandler {
+        override fun onPortConnected(port: Port) {
+            port.postMessage(JSONObject().put("type", "overrideFxAServer").put("url", serverConfig.contentUrl))
```
[the remark this answers is not in the document]
Comment (jackyzy823) [reply components/feature/accounts/src/main/java/mozilla/components/feature/accounts/FxaWebChannelFeature.kt]: updated.

**Comment by rfk** [comment]
(👋🏻 hi, I will try to take a look at this tomorrow, but flagging myself for review in the meantime just to make the status clear)

File: components/feature/accounts/src/main/assets/extensions/fxawebchannel/background.js [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+*/
+port.onMessage.addListener( event => {
+  if(event.type == "overrideFxAServer"){
+    browser.contentScripts.register({
+      "matches": [ event.url+"/*" ],
+      "js": [{file: "fxawebchannel.js"}],
+      "runAt": "document_start"
+    });
+    port.disconnect();
+  }
```
Comment (rfk) [inline components/feature/accounts/src/main/assets/extensions/fxawebchannel/background.js:20]: So, just to make sure I understand: this is a one-shot communication channel that lets the application signal to the extension, that it is using a custom FxA server at the given url. Is that right?

File: components/feature/accounts/src/main/assets/extensions/fxawebchannel/background.js [hunk 4]
```diff
@@ -0,0 +1,21 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/*
+Establish communication with native application.
+*/
+let port = browser.runtime.connectNative("mozacWebchannelBackground");
```
Comment (rfk) [inline components/feature/accounts/src/main/assets/extensions/fxawebchannel/background.js:8]: A purely stylistic suggestion, but could you please use the same constant name here as you do in the Kotlin code? Something like:

```
const WEB_CHANNEL_BACKGROUND_MESSAGING_ID = "mozacWebchannelBackground";
let port = browser.runtime.connectNative(WEB_CHANNEL_BACKGROUND_MESSAGING_ID);
```

The reason I ask is, I can see our future selves having to extend or debug something with webchannels in future, and trying to grep around in the codebase to remember where all the moving parts are. Making the different parts as similarly-shaped as possible may help that future exploration be more smooth.

**Review by rfk — APPROVED** [review]
Assuming I've correctly understood the intent here, this looks good to me and I believe it will correctly enforce the security properties we want. As long as it's not possible for a malicious extension to also call `browser.runtime.connectNative("mozacWebchannelBackground")`, but I believe we're already assuming that with the existing call to `browser.runtime.connectNative("mozacWebchannel")`, so...

File: components/feature/accounts/src/main/assets/extensions/fxawebchannel/background.js (hunk 3 above)
Comment (jackyzy823) [reply components/feature/accounts/src/main/assets/extensions/fxawebchannel/background.js:21]: yes , you're right.

File: components/feature/accounts/src/main/assets/extensions/fxawebchannel/background.js (hunk 4 above)
Comment (jackyzy823) [reply components/feature/accounts/src/main/assets/extensions/fxawebchannel/background.js]: done.

**Comment by jackyzy823** [comment]
It have been almost 5 months. And 

> Are there any chance to get this merged?
https://github.com/mozilla-mobile/android-components/issues/6225#issuecomment-884597830

Thanks.

**Comment by pocmo** [comment]
@grigoryk

**Comment by jackyzy823** [comment]
I don't want to be rude . but I feel hurt by @grigoryk keeping ignoring me.

So what do you **Mozillians** mean by
> We encourage you to participate in this open source project. We love pull requests, bug reports, ideas, (security) code reviews or any kind of positive contribution.

**Review by grigoryk — APPROVED** [review]
Hi, my apologies for the enormous delay. In a perfect world this would have merged a while back without needing my explicit r+ (reviews from Ryan and Christian were more than sufficient). But alas...

This looks fine to me after reading discussion in https://github.com/mozilla-mobile/android-components/issues/6225.

Thanks a bunch, and again, my apologies.

**Pull request merged**